packages/worker/src/worker-tuner.ts

```diff
@@ -141,11 +167,60 @@ const isResourceBasedTuner = (tuner: WorkerTuner): tuner is ResourceBasedTuner =
   Object.hasOwnProperty.call(tuner, 'tunerOptions');
 const isTunerHolder = (tuner: WorkerTuner): tuner is TunerHolder =>
   Object.hasOwnProperty.call(tuner, 'workflowTaskSlotSupplier');
-const isResourceBased = (sup: SlotSupplier): sup is ResourceBasedSlotsForType => sup.type === 'resource-based';
+const isResourceBased = (sup: SlotSupplier<any> | NativeSlotSupplier): sup is ResourceBasedSlotsForType =>
+  sup.type === 'resource-based';
+const isCustom = (sup: SlotSupplier<any> | NativeSlotSupplier): sup is CustomSlotSupplier<any> => sup.type === 'custom';
 
 type ActOrWorkflow = 'activity' | 'workflow';
+const abortString = '__ABORTED_BY_CORE__';
+
+class ErrorLoggingSlotSupplier<SI extends SlotInfo> implements CustomSlotSupplier<SI> {
+  readonly type = 'custom';
+
+  constructor(
+    private readonly supplier: CustomSlotSupplier<SI>,
+    private readonly logger: Logger
+  ) {}
+
+  async reserveSlot(ctx: SlotReserveContext, registerAbort: any): Promise<SlotPermit> {
+    const abortController = new AbortController();
+    registerAbort(() => abortController.abort(abortString));
+    return await this.supplier.reserveSlot(ctx, abortController.signal).catch((err) => {
```

Comment:
How bad will it be back in Rust world if the user function actually returned undefined or null rather than a real object? TS types are unfortunately not guarantees at runtime.

We usually don't get paranoid with these things in TS SDK, but since the contract is kind of inverted here, I'm wondering if we may need to be a tiny bit more careful.

Reply:
It shouldn't matter. Any returned value counts as a "permit" and gets used as the data for it. However I will double check that with a test. It's worth calling that out in the doc string too probably.

Reply:
If the rust-side don't ever look into the returned object, there me be absolutely problem here.

Reply:
Thank you! Fixed